Sources/NIOCore/ByteBuffer-core.swift

```diff
@@ -885,6 +886,30 @@ public struct ByteBuffer {
         return true
     }
 
+    /// The `ByteBuffer` will successfully be shrunk if the requested capacity is less than the current capacity,
+    /// and the requested capacity is more than the number of readable bytes in the buffer.
+    /// If either condition is not true, the buffer will not be shrunk.
+    ///
+    /// - Parameter desiredCapacity: The desired capacity for the buffers capacity to be shrunken to
+    /// - Returns: Bool indicating whether the buffer capacity has been shrunk to the desiredCapacity.
+    @inlinable
+    @discardableResult public mutating func shrinkBufferCapacity(to desiredCapacity: Int) -> Bool {
+        let desiredCapacity = ByteBuffer.addPaddingTo(desiredCapacity)
+        guard desiredCapacity < self.capacity, desiredCapacity > self.readableBytes else {
+            return false
+        }
+        self._copyStorageAndRebase(capacity: _toCapacity(desiredCapacity), resetIndices: true)
+        return true
+    }
+
+    /// Returns size of capacity with optimal padding
+    /// - Parameter initialCapacity: Capacity that needs expansion with padding
+    /// - Returns: Capacity with calculated padding
+    @inlinable
+    static func addPaddingTo(_ initialCapacity: Int) -> Int {
```

Comment:
```suggestion
    static func roundUpToUsableCapacity(_ initialCapacity: Int) -> Int {
```

I think we should be more specific here. You can also use this function elsewhere.

Reply:
Commited this change. And I can raise a follow up to use this across the repo.